New page: important (#15970)
mdn/content: files/en-us/web/css/important/index.md

---

> title: important

Author: ```suggestion
title: '!important'
```

tags:
  - CSS
  - Reference
  - CSS rule
  - important
  - \!important
# browser-compat: css.flag.important
---

The important flag, comprised of a `!` delimeter and `important` keyword, when added to a declaration in a selector block, makes the declaration an important declaration. 

## Description

To create an important declaration, add the "important flag", comprised of the delimiter token `!` followed by the `important` keyword, after the value in the declaration. While white space is allowed between the delimeter and the keyword, the important flag is generally written as `!important`, without any white space.

```css
selector {
  property: value; /* normal declaration */
  property: value !important; /* important declaration (standard) */
  property: value ! important; /* important declaration (non-standard) */
}
```

The important flag must come after the property value pair declaration value, preceded by at least one space. The `!important` or `! important` must be the last tokens in the declaration. In other words, there can be white space and comments between the flag and the declarations ending semi-colon, but nothing else.  

## Impact on the cascade

> When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed.  By default, declarations in an author (web developer) style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet. 

Reviewer: ```suggestion
When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. By default, declarations in a website author's style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet.
```

### Inline styles

When it comes to the [`style` attribute](/en-US/docs/Web/HTML/Global_attributes/style), inline normal styles take precedence over all normal declarations, no matter the origin and inline important styles take precedence over all other important author styles, no matter the layer, but are overridden by important user styles, important user-agent styles and transitions.

### !important and specificity

While `!important` is not part of determining specificity, it is related.  Important declarations override all other declarations from the same [origin and cascade layer](/en-US/docs/Web/CSS/Cascade). 

```css
#myElement#myElement#myElement .myClass.myClass p:hover {
  color: blue;
}

p {
  color: red !important;
}
```

> This is a case of over-specifying a selector. No matter how high the selector  [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precending.  In this case, the paragraph will always be red.

Reviewer: typo
```suggestion
This is a case of over-specifying a selector. No matter how high the selector [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precedence. In this case, the paragraph will always be red.
```

---
title: '!important'
slug: Web/CSS/important
tags:
  - CSS
  - Reference
  - CSS rule
  - important
  - \!important
# browser-compat: css.flag.important
---

> The important flag, comprised of a `!` delimiter and `important` keyword, when added to a declaration in a selector block, makes the declaration an important declaration. 

Reviewer 2: ```suggestion
A `!` delimiter followed by the `important` keyword sets the **important** flag on a declaration. The `!important` flag alters the rules selecting declarations inside the [cascade](/en-US/docs/Web/CSS/Cascade). A declaration that is not _important_ is called _normal_.
```

Author: still need to say "makes the declaration an important declaration."

Reviewer 2: Better?

```suggestion
A `!` delimiter followed by the `important` keyword marks the declaration as important. The `!important` flag alters the rules selecting declarations inside the [cascade](/en-US/docs/Web/CSS/Cascade). A declaration that is not _important_ is called _normal_.
```

> ## Description
> 

Reviewer 2: ```suggestion
```

> ## Description
> 

Reviewer 2: I don't think this section heading is necessary. We always start reference articles with a description.

```suggestion
```

> To create an important declaration, add the "important flag", comprised of the delimiter token `!` followed by the `important` keyword, after the value in the declaration. While white space is allowed between the delimeter and the keyword, the important flag is generally written as `!important`, without any white space.

Reviewer 2: ```suggestion
To mark a declaration important, add the _important flag_ (`!important`) after the value in the declaration. While white space is allowed between the delimiter and the keyword, the flag is generally written as `!important` without any white space.
```

```css
selector {
  property: value; /* normal declaration */
  property: value !important; /* important declaration (standard) */
  property: value ! important; /* important declaration (non-standard) */
}
```

> The important flag must come after the property value pair declaration value, preceded by at least one space. The `!important` or `! important` must be the last tokens in the declaration. In other words, there can be white space and comments between the flag and the declarations ending semi-colon, but nothing else.  

Reviewer 2: ```suggestion
The `!important` flag comes after the value of the property value pair declaration, preceded by at least one space. The `!important` flag must be the last token in the declaration. In other words, there can be white space and comments between the flag and the declaration's ending semi-colon, but nothing else.
```

## Impact on the cascade

> When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. By default, declarations in a website author's style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet.

Reviewer 2: ```suggestion
When it comes to `!important` declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. Without the important flag, declarations in the author's style sheets override declarations in a user's style sheet, which override declarations in the user-agent's default style sheet.
```

> When a declaration is important, the order of precedence is reversed. The user-agent important declarations override all user important declarations. All user important declarations override all important author declarations. And all important declarations take precedence over all animations.  

Reviewer 2: ```suggestion
When a declaration is `!important`, the order of precedence is reversed. Declarations marked as `!important` in the user-agent style sheets override all `!important` declarations in the user style sheets. Similarly, all `!important` declarations in the user style sheets override all `!important` declarations in the author's style sheets. Finally, all `!important` declarations take precedence over all animations.
```

> **Note:** All important declarations take precedence over all animations. `!important` is not valid on within [@keyframes animation](/en-US/docs/Web/CSS/@keyframes) declarations.

> Reversing the precedence order for important declarations ensures users with special needs, such as personalized color schemes or large fonts, can override author styles when needed with important declarations. It also ensures malicious extensions can't override important user-agent styles, which might break functionality or negatively impact security. 

Reviewer 2: ```suggestion
Reversing the precedence order for `!important` declarations ensures users with special needs, such as personalized color schemes or large fonts, can override author styles when needed by marking some declarations in their user's style sheet as `!important`. It also guarantees malicious extensions can't override `!important` user-agent styles, which might break functionality or negatively impact security.
```

> Does anything have precedence over important declarations? Yes. [Transitions](/en-US/docs/Web/CSS/CSS_Transitions). CSS transitions are a way to control the speed at which the property changes from one value to another.  A transitioning property, while transitioning from one value to another, will not match a specific important declaration. 

Reviewer 2: ```suggestion
Does anything have precedence over important declarations? Yes, [transitions](/en-US/docs/Web/CSS/CSS_Transitions). CSS transitions are a way to control the speed at which the property changes from one value to another. While transitioning from one value to another, a property will not match a specific important declaration.
```

```css
a {
  color: red !important;
  background-color: yellow;
  transition: all 2s linear;
}
a:hover {
  color: blue !important;
  background-color: orange !important;
}
```

> In this example, the color and background color will transition to the hovered state over two seconds. Even though the default states are normal and hover states are important declarations, the transition does happen.

Reviewer 2: ```suggestion
In this example, the `color` and `background-color` properties will transition to the hovered state over two seconds. Even though default states are normal declarations and hover states are `!important` declarations, the transition does happen.
```

### Cascade layers

> Within each of the three origins - author, user, and user-agent - normal declarations in unlayered styles override layered style declarations, with the last declared having precedence over the layers declared before it. The order of precedence is reversed for important declarations, with important declarations in the first declared layer taking precedence over important declarations in the last declared layer. All the important declarations have precedence over important declarations made outside of any layer.

Reviewer 2: ```suggestion
Within each of the three origins for style sheets – author, user, and user-agent – normal declarations in unlayered styles override layered style declarations, with the last declared having precedence over the layers declared before it. Important declarations reverse the order of precedence: `!important` declarations in the first layer take precedence over `!important!` declarations in the next layer, and so on. Also, all the `!important` declarations have precedence over `!important` declarations made outside of any layer.
```

### Inline styles

> When it comes to the [`style` attribute](/en-US/docs/Web/HTML/Global_attributes/style), inline normal styles take precedence over all normal declarations, no matter the origin and inline important styles take precedence over all other important author styles, no matter the layer, but are overridden by important user styles, important user-agent styles and transitions.

Reviewer 2: ```suggestion
Inline styles are styles defined using the [`style`](/en-US/docs/Web/HTML/Global_attributes/style) attributes. They can also be normal or important. Inline _normal_ styles take precedence over all _normal_ declarations, no matter the origin. Inline _important_ styles take precedence over all other _important_ author styles, no matter the layer, but important styles from user's or user-agent's style sheets and transitions override them.
```

### !important and specificity

> While `!important` is not part of determining specificity, it is related.  Important declarations override all other declarations from the same [origin and cascade layer](/en-US/docs/Web/CSS/Cascade). 

Reviewer 2: ```suggestion
While `!important` is not part of determining specificity, it is related. Important declarations override all other declarations from the same [origin and cascade layer](/en-US/docs/Web/CSS/Cascade).
```

```css
#myElement#myElement#myElement .myClass.myClass p:hover {
  color: blue;
}

p {
  color: red !important;
}
```

> This is a case of over-specifying a selector. No matter how high the selector  [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precending.  In this case, the paragraph will always be red.

Reviewer 2: ```suggestion
This example displays a case of over-specifying a selector. No matter how high the selector [specificity](/en-US/docs/Web/CSS/Specificity) matches a normal declaration, an `!important` declaration from the same source and cascade layer will always have precedence. In this case, the paragraph will always be red.
```

> When two important declarations are applied to the same element, with the same origin and layer, the declaration with a greater specificity will be applied.

Reviewer 2: ```suggestion
When two `!important` declarations from the same origin and layer apply to the same element, browsers select and use the declaration with the highest specificity.
```

```css
#myElement p {
    color: green !important;
}

p {
    color: purple !important;
}
```

In this case, the selector specificity matters. Only if the selectors had the same specificity would source order matter.

## Impact on shorthand properties

> Declaring a shorthand property to be `!important` sets all of sub-properties `!important`. To two following selector style blocks are equivalent:

Reviewer 2: ```suggestion
Declaring a shorthand property to be `!important` defines all of the longhand properties as`!important`. To two following selector style blocks are equivalent:
```

```css
p {
  background: blue !important;
}

p {
  background-image: none !important;
  background-position: 0% 0% !important;
  background-size: auto auto !important;
  background-repeat: repeat !important;
  background-origin: padding-box !important;
  background-clip: border-box !important;
  background-attachment: scroll !important;
  background-color: blue !important;
}
 ``` 

> This is one of the several reasons avoiding `!important` is generally recommended.

Reviewer 2: ```suggestion
This example shows one of the several reasons avoiding `!important` is generally recommended.
```

## Best practices

> Avoid using `!important` to override specificity. When intentionally creating important declarations for UI requirements, comment your CSS to explain to maintainers why a feature must not be overridden.

Reviewer 2: ```suggestion
Avoid using `!important` to override specificity. When intentionally creating `!important` declarations for UI requirements, comment in your CSS code to explain to maintainers why they should not override that feature.
```

> If working to override styles with high scecificity that are not under your control, such as styles in a 3rd party plugin declared with an [id selector](/en-US/docs/Web/CSS/ID_selectors), consider importing the 3rd party stylesheet script into a [named or anonymous layer](/en-US/docs/Web/CSS/@layer) as your first cascade layer, instead of using `!important`. As long as the external styles do not include important declarations, your styles will take precedence over the widget styles no matter the specificity. 

Reviewer 2: ```suggestion
Even when working to override high-specificity styles not under your control, such as styles in a 3rd party plugin declared with an [id selector](/en-US/docs/Web/CSS/ID_selectors), you don't need to use `!important`. Consider instead importing the 3rd party stylesheet script into a [named or anonymous layer](/en-US/docs/Web/CSS/@layer) as your first cascade layer, instead of using `!important`. As long as the external styles do not include `!important` declarations, your styles will take precedence over the widget styles, no matter the specificity.
```

> If needing to override external stylesheet containing important declarations, create a cascade layer of importnant overrides and declare that layer first. 

Reviewer 2: ```suggestion
If you need to override an external stylesheet containing `!important` declarations, create a cascade layer containing the needed overrides, and declare that layer first.
```

### Accessibility

> Important styles from a user stylesheet take precedence over author important declarations, meaning adding an important flag to a site's styles will not prevent individual users with special requirements, such as large fonts, from being able to override your styles with their own important styles.

Reviewer 2: ```suggestion
Important styles from a user stylesheet take precedence over the author style sheet's `!important` declarations, meaning adding an `!important` flag to a site's styles will not prevent individual users with special requirements, such as large fonts, from being able to override your styles by adding`!important` styles in their own user's style sheet.
```

---

> title: important

Reviewer 2: Drive-by comment:

Shouldn't we add the bang to the title? `!important`

Author: i was going to ask "how do you do that". If I include the ! I get an error.

Need to know:
* which, if any,  {{}} should be included for side nav.
* is there a way to change the URL and title to !important?

Reviewer 3: @estelle You might try quoting the title (untested): `title: '!important'`

No idea about the url.

Reviewer: > change the URL to !important

`!` is a valid character in URLs and file names
